pkgs/http/lib/src/base_request.dart

```diff
@@ -89,14 +89,33 @@ abstract class BaseRequest {
   bool _finalized = false;
 
   static final _tokenRE = RegExp(r"^[\w!#%&'*+\-.^`|~]+$");
+
   static String _validateMethod(String method) {
     if (!_tokenRE.hasMatch(method)) {
       throw ArgumentError.value(method, 'method', 'Not a valid method');
     }
     return method;
   }
 
-  BaseRequest(String method, this.url)
+  /// On upload progress callback.
+  ///
+  /// If defined, this callback will be called when the upload progress changes.
+  ///
+  /// In browser, uses XMLHttpRequest's "xhr.upload.onLoad" event.
+  ///
+  /// In IO, uses the yield length of the stream. The total length of the bytes
+  /// yielded by the stream at any given moment is "uploaded" and the total
```

Comment:
@brianquinlan, @lrhn - how meaningful do we expect this to be? Is watching a transform step in a stream being read by `dart:io` HTTP handling an accurate way to understand upload progress?

Reply:
Not exactly.  While I was writing the altogic_dart package, I duplicated it.  Meanwhile, I split all yields larger than the specified chunk size according to this size.  But adding this to the official package would have caused more confusion. Currently MultipartRequest splits into specific chunks and progress works correctly as the part is actually.  But it needs to be divided into more accurate parts for big files. 

E.g current results like this for big files : 0.1- 1-99.3-100

If large yields are divided, loading can be done with closer ratios.

Reply:
There is two methods for this. 

1) Like browser client, read stream until done, split by defined chunk size. This may cause more memory consumption. 

2) Split big parts by chunk size in the ByteStream. But this splitting will be done  for browser also and it unnecessary.  Because browser client works different. 

If there is a solution you approve, I can code it.  I also need to find out if the chuck size will be a static value or a parameter.  Parameters can cause confusion.  if it is static how much should it be?

Reply:
Or make `total` nullable and don't provide it when you don't have one.

If the `loaded` is defined as being bytes, at least when total is `null`, then the user can still be shown a meaningful progress, it'll just be "49 MB" instead of "37%".


I'd never read the entire stream and then chunk it afterwards. That's just useless overhead. If everything comes in one big chunk, that's just what you have to deal with.

Reply:
Making `total` nullable does not fix the problem. Instead, the `ByteStream .fromBytes` consturctor simply splits the List<int>(bytes) into parts.

If MultipartFile is created with a stream, meaningful data is now obtained. But when created with fromBytes it pipes it as one piece. This prevents the rates from becoming meaningful. 

Also at the moment onUploadProgress is not called with a rate anyway. There are loaded and total. Since the callback is triggered only where the total is calculated, the developer can calculate the rate.

Reply:
> @brianquinlan, @lrhn - how meaningful do we expect this to be? Is watching a transform step in a stream being read by `dart:io` HTTP handling an accurate way to understand upload progress?

Probably not. When you `write` to a dart socket it doesn't mean that the data has actually left the local machine (or even been copied to OS memory).

Reply:
Of the 6 implementations that I know about, you could get good numbers for:
- browser
- cupertino_http

and bad numbers for:
- fetch_client (though this should improve with time as the JavaScript streaming API improves)
- cronet_http
- io
- java_http (not sure about this, based on Java HttpURLConnection)